### ISSUE TITLE
Moodle migration from AWS EC2 instance to AWS ECS by using Docker Container

### ISSUE THREAD

**Issue by johnysd-GR** [bitnami/bitnami-docker-moodle#96]
Hello
I'd like to know how to migrate my Bitnami Moodle instance from AWS EC2 to AWS ECS by using docker container image and furthermore how to maintain the latest updates on my new environment?

I'm now running the following versions
Moodle : 3.4.1
PHP : 7.0.26
MySQL : 5.7.22 (source)
MySQL : 5.7.25 (destination)

History:

I've tried many different approaches but I was always hitting on a wall.
Once I was not running the exact MySQL version, the next time the PHP was not compatible with the files imported, the next time the database update process was stuck forever etc. 
The last time that I tried, I was asked to run the initial script setup on every container start with no way to avoid it even if I had data to my moodledata folder and my DB as well. Every time the container was starting up was halted to the message that the database is not empty.
I tried to create a script and use a temporary schema just for this initial phase and after the initial build of the schema and the tables to switch to the productive DB.
_This DB initialization issue has been fixed by adding the MOODLE_SKIP_INSTALL flag._

I've tried both https://github.com/moodlehq/moodle-docker and https://hub.docker.com/r/bitnami/moodle/ docker images with no success

### **That was my initial post to Bitnami Community but seems that my issue is more related to this GitHub thread as i'm trying to use this docker image**

As per Bitnami-Docker-Moodle issue, [#95](https://github.com/bitnami/bitnami-docker-moodle/issues/95) regarding the initial installation seems that have been fixed as Wesley Salazar reports and I can confirm.

But, I'm in the same problem loop again.
I'm trying to run a copy of my production Moodle installation locally by following the steps below.

I used [this](https://github.com/bitnami/bitnami-docker-moodle) bitnami docker moodle image for my tests

Step1: I'm creating the MariaDB and Moodle containers by docker-compose file and everything works fine.
Step2: Trying to change the docker volumes with persistent folders locally and works with some page load delays.
Step3: Paste my moodle folder content (from my production env) into the moodle folder created by step 2 into this "data persistent" folder which I have already clear, set the "MOODLE_SKIP_INSTALL=yes" variable, change my config.php file in order to be connected to my local MySQL DB (not the containerised MariaDB) and bring the containers up again.

At this point, I realized that the structure is different and there is an init script running on Moodle Installation step which creates folders into moodle directory (even that I get the message "The installation wizard will be skipped and parameters for configuring the user account and SMTP credentials will be ignored.") 
_FYI: My productive moodle folder contains only .bitnamimeta,backup,conf,htdocs,licenses,moodledata folders and bnconfig and updateip files, but the script creates at least 50 more folders and 40 more files in there._

Furthermore:
1. The connection with my dedicated local DB seems to work (mysql-c INFO  MySQL server listening and working at 192.168.1.3:3306) but I can't log in with an existing user (confirmed that exists in this DB and the password is correct).
2. The pages can't be loaded correctly and seem like PHP issue, the images are missing as well and some js scripts are failed to load.

My expectations from the moodle container image was to start the container by indicating the DB variables and the location of the moodledata directory by binding a local folder as volume and get the Moodle service up and running.

**Comment by marcosbc** [comment]
Hi @johnysd-GR,


>FYI: My productive moodle folder contains only .bitnamimeta,backup,conf,htdocs,licenses,moodledata folders and bnconfig and updateip files, but the script creates at least 50 more folders and 40 more files in there.

The Bitnami stacks have a different folder organization in the Moodle installation directory. In order to migrate the AWS installation to the container, you need to copy the following directories from the stack to the container's `/bitnami/moodle`:

- `/opt/bitnami/apps/moodle/moodledata`
- `/opt/bitnami/apps/moodle/htdocs/*` (all files and directories)

Afterwards you need to update Moodle's `config.php` file and fix the value of `$CFG->dataroot` so it points to `/bitnami/moodle/moodledata` and the database credentials.

In addition to that, you need to ensure the file `/bitnami/moodle/.initialized` exists so the entrypoint doesn't override the files you copied.

**Comment by johnysd-GR** [comment]
Hi @marcosbc 
Thank you for your helpful feedback.
I get over a major issue but I'm still in the loop.

Since your answer, I'm trying to create an environment which meets my productive environment's versions and dependencies (which are Moodle 3.7.1 - MySQL 5.7.22 - PHP 7.0.26).
For this purpose, I did the following.

1. I cleared the docker-compose.yml file by deleting the mariadb service part in order to use the Moodle container independent (docker-compose is not available on AWS platform).
2. I used the image _'bitnami/moodle:3.4.1-r0'_ in order to meet my productive Moodle version
3. I created a new container with MySQL 5.7.22 in order to import and host my productive data

and now the issue...

By changing the "config.php" and after I made the changes to point at my newly created schema, I didn't manage to log in with my production user. I did not even manage to login with default moodle "user:bitnami" user. I can see into the database that the user that I'm trying to use exists.
May it be related to some password hashing? 

I have to mention that on my last successfully `docker-compose up` command, it took over 20 mins to go from line `mysql-c INFO MySQL server listening and working` to `moodle  INFO  Configuring Cron Jobs...` which stayed more 5 mins. After that, I checked my DB and it was shrunk from 177MB to 158MB. Of course I tried to login imediately after this initial start with no success and after re-import my productive data to this DB with no success (*after re-import the data size is 175MB).

Bellow the log output:

> Creating network "bitnami-docker-moodle-master-moodle341-mysql5722_default" with the default driver
Creating bitnami-docker-moodle-master-moodle341-mysql5722_moodle_1 ... done
Attaching to bitnami-docker-moodle-master-moodle341-mysql5722_moodle_1
moodle_1  |
moodle_1  | Welcome to the Bitnami moodle container
moodle_1  | Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-moodle
moodle_1  | Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-moodle/issues
moodle_1  | Send us your feedback at containers@bitnami.com
moodle_1  |
moodle_1  | nami    INFO  Initializing apache
moodle_1  | apache  INFO  ==> Reconfiguring PID file location...
moodle_1  | nami    INFO  apache successfully initialized
moodle_1  | nami    INFO  Initializing php
moodle_1  | nami    INFO  php successfully initialized
moodle_1  | nami    INFO  Initializing moodle
moodle_1  | mysql-c INFO  Trying to connect to MySQL server
moodle_1  | mysql-c INFO  Found MySQL server listening at 192.168.88.254:3304
moodle_1  | mysql-c INFO  MySQL server listening and working at 192.168.88.254:3304
moodle_1  | moodle  INFO  Configuring Cron Jobs...
moodle_1  | moodle  INFO
moodle_1  | moodle  INFO  ########################################################################
moodle_1  | moodle  INFO   Installation parameters for moodle:
moodle_1  | moodle  INFO     Persisted data and properties have been restored.
moodle_1  | moodle  INFO     Any input specified will not take effect.
moodle_1  | moodle  INFO   This installation requires no credentials.
moodle_1  | moodle  INFO  ########################################################################
moodle_1  | moodle  INFO
moodle_1  | nami    INFO  moodle successfully initialized
moodle_1  | INFO  ==> Starting moodle...
moodle_1  |
moodle_1  | 192.168.16.1 - - [27/Aug/2019:13:06:50 +0000] "GET / HTTP/1.1" 200 11368

Thank you in advance!

**Comment by marcosbc** [comment]
Hi @johnysd-GR, unfortunately `bitnami/moodle:3.4.1-r0` is out-of-date and we're unable to provide new features for old Moodle releases.

If the version were more recent, we would be able to give some instructions to use that version. However it is more than a year old and many major features have been implemented for the different components of the container (not only Moodle but also Apache and PHP).

The only way that I can think of getting this to work, is by manually overwriting all 3.7 files in an already initialized Moodle container by version 3.4 files. After that create the `config.php` with proper DB configuration (it should be the same configuration as the one in your env. vars), and re-create the containers:

    docker-compose down && docker-compose up

If it worked, it should have recreated the config.php via `install.php` without recreating database tables. However, unfortunately we cannot guarantee this will work.

Our suggestion is that you instead upgrade your Moodle 3.4.1 external database to 3.7.1 and then use the latest release of the container, to avoid any issues.

**Comment by johnysd-GR** [comment]
Hello.

I was preparing yesterday, before your answer, to send you the following feedback.
I will post all these here just for information of anyone else which is facing the same case as these are from my attempts to use a docker container to migrate my EC2 installation to the locally created containerized docker environment.


The problem was the following:

I saw that the Moodle version was running back at version 3.4.1 after following your steps. Like some files which I use from my productive htdocs folder define the Moodle version and change it from 3.7.1.

So,
1. If I run the yml file with no persistent data and an empty DB `MOODLE_SKIP_INSTALL=no`, Moodle is functional on the latest version 3.7.1.
2. If I run the above and change after the initial setup the initialized moodle folder with my production moodle folder (without importing my DB data), I can login with user "user:bitnami" only and the site is not functional (theme does not load correctly, some data records in database tables can not be found on administration etc).
3. If I run it with only the DB data `MOODLE_SKIP_INSTALL=yes` and no moodle folder existence, I get the following showstopper message which is contradictory as it stacks to some tables after the recognition of an already initialized DB and the "The installation wizard will be skipped" line. 
*I was expecting the DB upgrade page and the upgrade of the existing DB (3.4.1) to the version just installed (3.7.1).
>Creating network "bitnami-docker-moodle-master-latest_default" with the default driver
Creating bitnami-docker-moodle-master-latest_moodle_1 ... done
Attaching to bitnami-docker-moodle-master-latest_moodle_1
moodle_1  |
moodle_1  | Welcome to the Bitnami moodle container
moodle_1  | Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-moodle
moodle_1  | Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-moodle/issues
moodle_1  |
moodle_1  | nami    INFO  Initializing apache
moodle_1  | nami    INFO  apache successfully initialized
moodle_1  | nami    INFO  Initializing php
moodle_1  | nami    INFO  php successfully initialized
moodle_1  | nami    INFO  Initializing mysql-client
moodle_1  | nami    INFO  mysql-client successfully initialized
moodle_1  | nami    INFO  Initializing moodle
moodle_1  | mysql-c INFO  Trying to connect to MySQL server
moodle_1  | mysql-c INFO  Found MySQL server listening at 192.168.88.254:3304
moodle_1  | mysql-c INFO  MySQL server listening and working at 192.168.88.254:3304
moodle_1  | moodle  INFO  Running Moodle install. Please be patient...
moodle_1  | moodle  INFO  You provided an already initialized Moodle database.
moodle_1  | moodle  INFO  The installation wizard will be skipped and parameters for configuring the user account and SMTP credentials will be ignored.
moodle_1  | Error executing 'postInstallation': Default exception handler: Table "badge_alignment" already exists Debug: can not rename table
moodle_1  | Error code: ddltablealreadyexists
moodle_1  | * line 512 of /lib/ddl/database_manager.php: ddl_exception thrown
moodle_1  | * line 2869 of /lib/db/upgrade.php: call to database_manager->rename_table()
moodle_1  | * line 1809 of /lib/upgradelib.php: call to xmldb_main_upgrade()
moodle_1  | * line 180 of /admin/cli/upgrade.php: call to upgrade_core()
moodle_1  |
moodle_1  |
bitnami-docker-moodle-master-latest_moodle_1 exited with code 1



Today status, after your suggestion to upgrade my existing installation before migrate to containers is the following.

1. I've created a new EC2 instance as Development environment (we haven't this need before) by using an AMI of the productive EC2 instance in order to use the data and to be in the same version lvl as my productive instance.
2. I'm following the steps from [this bitnami docs thread](https://docs.bitnami.com/azure/apps/moodle/administration/upgrade/) even if it is referring to as "for azure"
3. I'm decided to proceed with the "step by step" approach which means upgrade from version 3.4.1 to 3.5 first by using the moodle-latest-35.tgz file on the "Download the new Moodle Version" step. After this upgrade, I'll continue to the next stable version.
4. All steps ran smoothly and I reach the final step "Navigate to the application from a browser and follow the steps to upgrade the database to the latest version.". 
When I open my application from my browser, everything seems to work like before except the login function. I have the known case where none of the known users are recognized and the error message is "Invalid login, please try again" (the config.php file is a copy from my htdocs.backup folder which is the one worked before).

Shouldn't I expect the DB upgrade screen after the upgrade steps?

PS: I had an idea today, to start the functional docker environment in version 3.4.1 with my persistent folders and DBs, and start upgrading the environment by changing the Moodle stable version images on my yml file.

**Comment by marcosbc** [comment]
Hi @johnysd-GR, the DB upgrade screen should have appeared. In any case it is possible for you to do this manually by running these commands:

    cd /opt/bitnami/moodle
    /opt/bitnami/php/bin/php admin/cli/upgrade.php --non-interactive --allow-unstable

You can also stop and start the container, as this command will be executed automatically at startup. But it's probably best you do it running in case you see any error (and the container does not end up crashing).

**Comment by johnysd-GR** [comment]
Hi
After my ECS test failure with the following logs 
```
[0m
[0m[1mWelcome to the Bitnami moodle container[0m
[0mSubscribe to project updates by watching [1mhttps://github.com/bitnami/bitnami-docker-moodle[0m
[0mSubmit issues and feature requests at [1mhttps://github.com/bitnami/bitnami-docker-moodle/issues[0m
[0m
nami INFO Initializing apache
nami INFO apache successfully initialized
nami INFO Initializing php
nami INFO php successfully initialized
nami INFO Initializing mysql-client
nami INFO mysql-client successfully initialized
nami INFO Initializing moodle
mysql-c INFO Trying to connect to MySQL server
mysql-c INFO Found MySQL server listening at dqfdev.cj31ql4bbkes.eu-west-1.rds.amazonaws.com:3306
mysql-c INFO MySQL server listening and working at dqfdev.cj31ql4bbkes.eu-west-1.rds.amazonaws.com:3306
moodle INFO Running Moodle install. Please be patient...
Error executing 'postInstallation': Program exited with exit code 1
```

I came back to my local environment and start tests again.

1. I copied my files (as you @marcosbc mentioned) and reinsert my db data. Started my already existing container and everything ran smooth and I was able to login with my user as well.
2. When I simply ran ```docker-compose down && docker-compose up``` I got the bellow logs and the message on my browser which was informing me about something which I didn't copy at that time (it was a message like the ones appeared with the orange frame in the middle of a white screen)

```
moodle_1  |
moodle_1  | Welcome to the Bitnami moodle container
moodle_1  | Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-moodle
moodle_1  | Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-moodle/issues
moodle_1  |
moodle_1  | nami    INFO  Initializing apache
moodle_1  | apache  WARN  The directory '/bitnami/apache/conf' was externally mounted. This is a legacy configuration and will be deprecated soon. Please mount configuration files at '/opt/bitnami/apache/conf' instead. Find an example at: https://github.com/bitnami/bitnami-docker-apache#full-configuration
moodle_1  | apache  WARN  Restoring configuration at '/bitnami/apache/conf' to '/opt/bitnami/apache/conf'...
moodle_1  | nami    INFO  apache successfully initialized
moodle_1  | nami    INFO  Initializing php
moodle_1  | php     WARN  The directory '/bitnami/php/conf' was externally mounted. This is a legacy configuration and will be deprecated soon. Please mount configuration files at '/opt/bitnami/php/conf' instead.
moodle_1  | php     WARN  Restoring configuration at '/bitnami/php/conf' to '/opt/bitnami/php/conf'...
moodle_1  | nami    INFO  php successfully initialized
moodle_1  | nami    INFO  Initializing mysql-client
moodle_1  | nami    INFO  mysql-client successfully initialized
moodle_1  | nami    INFO  Initializing moodle
moodle_1  | mysql-c INFO  Trying to connect to MySQL server
moodle_1  | mysql-c INFO  Found MySQL server listening at 192.168.88.254:3304
moodle_1  | mysql-c INFO  MySQL server listening and working at 192.168.88.254:3304
moodle_1  | moodle  INFO  Running Moodle install. Please be patient...
moodle_1  | moodle  INFO  You provided an already initialized Moodle database.
moodle_1  | moodle  INFO  The installation wizard will be skipped and parameters for configuring the user account and SMTP credentials will be ignored.
moodle_1  | moodle  INFO  Configuring Cron Jobs...
moodle_1  | moodle  INFO
moodle_1  | moodle  INFO  ########################################################################
moodle_1  | moodle  INFO   Installation parameters for moodle:
moodle_1  | moodle  INFO     Username: user
moodle_1  | moodle  INFO     Password: **********
moodle_1  | moodle  INFO     Email: user@example.com
moodle_1  | moodle  INFO   (Passwords are not shown for security reasons)
moodle_1  | moodle  INFO  ########################################################################
moodle_1  | moodle  INFO
moodle_1  | nami    INFO  moodle successfully initialized
moodle_1  | INFO  ==> Starting moodle...
moodle_1  | Starting Apache...
moodle_1  | [Fri Sep 06 13:47:30.454594 2019] [ssl:warn] [pid 93] AH01909: localhost:443:0 server certificate does NOT include an ID which matches the server name
moodle_1  | [Fri Sep 06 13:47:30.465170 2019] [ssl:warn] [pid 93] AH01909: localhost:443:0 server certificate does NOT include an ID which matches the server name
moodle_1  | [Fri Sep 06 13:47:30.527716 2019] [ssl:warn] [pid 93] AH01909: localhost:443:0 server certificate does NOT include an ID which matches the server name
moodle_1  | [Fri Sep 06 13:47:30.536687 2019] [ssl:warn] [pid 93] AH01909: localhost:443:0 server certificate does NOT include an ID which matches the server name
moodle_1  | [Fri Sep 06 13:47:30.649676 2019] [mpm_prefork:notice] [pid 93] AH00163: Apache/2.4.41 (Unix) OpenSSL/1.1.0k PHP/7.3.8 configured -- resuming normal operations
moodle_1  | [Fri Sep 06 13:47:30.650115 2019] [core:notice] [pid 93] AH00094: Command line: 'httpd -f /opt/bitnami/apache/conf/httpd.conf -D FOREGROUND'
moodle_1  | 172.19.0.1 - - [06/Sep/2019:13:47:33 +0000] "GET /admin/environment.php HTTP/1.1" 503 544
moodle_1  | 172.19.0.1 - - [06/Sep/2019:13:47:45 +0000] "GET /favicon.ico HTTP/1.1" 404 196
```
PS: It took 3 minutes to reach the line "_The installation wizard will be skipped and parameters for configuring the user account and SMTP credentials will be ignored._" and 2 more minutes on "_Configuring Cron Jobs..._"

3. Then I ran again the `docker-compose down && docker-compose up` and I got this on my logs...


```
moodle_1  |
moodle_1  | Welcome to the Bitnami moodle container
moodle_1  | Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-moodle
moodle_1  | Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-moodle/issues
moodle_1  |
moodle_1  | nami    INFO  Initializing apache
moodle_1  | apache  WARN  The directory '/bitnami/apache/conf' was externally mounted. This is a legacy configuration and will be deprecated soon. Please mount configuration files at '/opt/bitnami/apache/conf' instead. Find an example at: https://github.com/bitnami/bitnami-docker-apache#full-configuration
moodle_1  | apache  WARN  Restoring configuration at '/bitnami/apache/conf' to '/opt/bitnami/apache/conf'...
moodle_1  | nami    INFO  apache successfully initialized
moodle_1  | nami    INFO  Initializing php
moodle_1  | php     WARN  The directory '/bitnami/php/conf' was externally mounted. This is a legacy configuration and will be deprecated soon. Please mount configuration files at '/opt/bitnami/php/conf' instead.
moodle_1  | php     WARN  Restoring configuration at '/bitnami/php/conf' to '/opt/bitnami/php/conf'...
moodle_1  | nami    INFO  php successfully initialized
moodle_1  | nami    INFO  Initializing mysql-client
moodle_1  | nami    INFO  mysql-client successfully initialized
moodle_1  | nami    INFO  Initializing moodle
moodle_1  | mysql-c INFO  Trying to connect to MySQL server
moodle_1  | mysql-c INFO  Found MySQL server listening at 192.168.88.254:3304
moodle_1  | mysql-c INFO  MySQL server listening and working at 192.168.88.254:3304
moodle_1  | Error executing 'postInstallation': ERROR!!!  The code you are using is OLDER than the version that made these databases!
moodle_1  |
moodle_1  |
bitnami-docker-moodle-master-latest_moodle_1 exited with code 1
```
4. Copy again the files, restart the container and got this message but this time on my browser
```
ERROR!!! The code you are using is OLDER than the version that made these databases!
```
Press the "_Continue_" button and nothing happened.

5. At this point, I decided to just delete all the existing moodle files and copy again mine, which worked including the login function.

From my experience so far, i can understand that there is something on the init script changes my files even if I put the variable `MOODLE_SKIP_INSTALL=yes`. The message with the username and the password after the message that "the installation wizard will be skipped" and the "configuration of the cron jobs" seems the wrong part of the init script.

 

I will fight with ECS next week and keep you posted...

To Be Continued

**Comment by johnysd-GR** [comment]
Hello @marcosbc 
I've spent so much time on this and other projects and tasks are running as well.
I'd like to paste my functional local docker configuration here

> version: '2'
services:
  moodle:
    image: 'bitnami/moodle:3.4.1-r4'
    #image: 'bitnami/3.7.2-ol-7-r5'
    environment:
      - MARIADB_HOST=my DB host
      - MARIADB_PORT_NUMBER=3304
      - MOODLE_DATABASE_NAME=moodle
      - MOODLE_DATABASE_USER=moodle
      - MOODLE_DATABASE_PASSWORD=password
      - MOODLE_SKIP_INSTALL=yes
    ports:
      - '80:80'
      - '443:443'
    volumes:
      - '/Users/myuser/Documents/dockerUAT/bitnami-docker-moodle-master-latest/data:/bitnami'

The target now is just to migrate the current Moodle installation to ECS as it is (not upgrade*) in order to shut down the dedicated instance running for my company's e-learning platform and of course reduce the costs.

*After my experience with this task, I don't feel confident enough to do the upgrade to our current production installation (EC2 dedicated instance with Moodle 3.4.1/PHP 7.0.26/MySQL 5.7.26)

I don't know when I'll have time to work on this task again but this is a "must" task which has been just placed to "on hold" list and I will face it again soon.

**Comment by marcosbc** [comment]
Thanks for letting us know.

Our recomendation is for you to focus on the Moodle upgrade in your production installation, and afterwards move to ECS. This is because of the required features that were implemented recently to the Moodle container in newer releases.

The only issue we see in your production server is upgrading PHP to 7.1. Once that is done you should not find any issues to upgrade Moodle to >=3.7.1 in that server. If you want to pursue this route you could follow the steps below:

- Backup your instance to avoid any loss of data in case of failure.
- Install the latest LAMP >=7.1 like mentioned [in this post](https://community.bitnami.com/t/upgrade-to-the-latest-php-or-the-latest-lamp-stack-within-the-lightsail-aws-service/57019/2).
- Stop any running services (PHP-FPM, MySQL, Apache).
- Replace Apache and PHP configuration files (in `apache2/conf` and `php/etc` respectively) in the new installation, with the ones from the old installation.
- Copy Moodle from the old installation in `apps/moodle`, preserving same file/directory permissions (i.e. `sudo cp -rp /opt/bitnami.back/apps/moodle /opt/bitnami/apps/`).
- The MySQL database does not need to be copied, meaning for now you just replace the new MySQL with the old (i.e. `sudo rm -rf /opt/bitnami/mysql && cp -rp /opt/bitnami/mysql /opt/bitnami`).
- Start services again.

**Comment by johnysd-GR** [comment]
Thank you very much for your support @marcosbc 
My current production installation uses a Moodle Certified by Bitnami AMI and an RDS instance for MySQL. So MySQL is not an issue as it is a standalone instance and can be managed separately (except it is?!).
I 'm thinking to use the latest available AMI to create a separate EC2-UAT environment and copy my files from the productive one there. This could easily be my playground for the upgrade task.
After my tests/tries and the successful upgrade, I could use this UAT instance until the ECS migration and close the outdated one.
My only concern is the difference in versions that the old AMI and the new one may have.

**Comment by marcosbc** [comment]
That's fine. If you launch the new Bitnami Moodle stack with PHP 7.1, it should be enough with moving all of your current Moodle files to the new instance's `/opt/bitnami/apps/moodle` directory to replicate the existing structure.

Once that is done, you should be able to start the upgrade process to 3.7.x.
